### PR TITLE
fix($httpParamSerializerJQLike): call functions as jQuery does

### LINKED ISSUE #16138: fix($httpParamSerializerJQLike): call functions as jQuery does
**I'm submitting a ...**

- [x] bug report
- [ ] feature request
- [ ] other (Please do not submit support requests here (see above))

**Current behavior:**

Function properties are ignored by `$httpParamSerializerJQLike()`.

**Expected / new behavior:**

Functions should be called and the return value used, the same way [jQuery does][1].

[1]: https://github.com/jquery/jquery/blob/a6b0705294d336ae2f63f7276de0da1195495363/src/serialize.js#L63-L65

### PR DESCRIPTION
**Pull request by frederikprijck**
Closes #16138

**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Bug fix

**What is the current behavior? (You can also link to an open issue here)**
Previously, `httpParamSerializerJQLike` did not handle properties of type function.

**What is the new behavior (if this is a feature change)?**
This commit ensures function properties are executed and the return value is used.

**Does this PR introduce a breaking change?**
See: https://github.com/angular/angular.js/issues/16138#issuecomment-319066731

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

### CODE REVIEW

File: src/ng/http.js [hunk 1]
```diff
@@ -122,8 +122,12 @@ function $HttpParamSerializerJQLikeProvider() {
                 (topLevel ? '' : ']'));
           });
         } else {
-          parts.push(encodeUriQuery(prefix) + '=' +
-              (toSerialize == null ? '' : encodeUriQuery(serializeValue(toSerialize))));
+          if (isFunction(toSerialize)) {
```
Comment (gkalpak) [inline src/ng/http.js]: Merge with the `else` above for consistency:

```js
  ...
} else if (isFunction(toSerialize)) {
  ...
} else {
  ...
}
```

File: test/ng/httpSpec.js [hunk 2]
```diff
@@ -2369,12 +2369,36 @@ describe('$http param serializers', function() {
...
+      expect(jqrSer({a: function() { return 'b'; }, foo: {'bar': 'barv', 'baz': 'bazv'}})).toEqual(
+        'a=b&foo%5Bbar%5D=barv&foo%5Bbaz%5D=bazv');
+        //a=b&foo[bar]=barv&foo[baz]=bazv
+    });
+
+    it('should serialize objects with function properties returning an object by repeating param name with [key] suffix', function() {
+      expect(jqrSer({a: function() { return { b: 'c' }; }, foo: {'bar': 'barv', 'baz': 'bazv'}})).toEqual(
+        'a%5Bb%5D=c&foo%5Bbar%5D=barv&foo%5Bbaz%5D=bazv');
+        //a[b]=c&foo[bar]=barv&foo[baz]=bazv
+    });
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: These two tests are testing more (independent) things than necessary. They should focus on the `function` part. The rest is irrelevant.

File: test/ng/httpSpec.js [hunk 3]
```diff
@@ -2369,12 +2369,36 @@ describe('$http param serializers', function() {
...
          //a[]=b&a[1][c]=d&e[f]=g&e[h][]=i&e[h][]=j
     });
 
+    it('should serialize nested objects with function properties by repeating param name with [key] suffix', function() {
+      expect(jqrSer({a: 'b' , foo: {'bar': function() { return 'barv'; }, 'baz': 'bazv'}})).toEqual(
+        'a=b&foo%5Bbar%5D=barv&foo%5Bbaz%5D=bazv');
+        //a=b&foo[bar]=barv&foo[baz]=bazv
+    });
+
+    it('should serialize nested objects with function properties returning an object by repeating param name with [key] suffix', function() {
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: 👍

**Review by gkalpak — CHANGES_REQUESTED** [review]
There are no tests with functions inside arrays 😞 

Can you also change the commit message to make it more clear that previously function were stringified (not ignored). This will make it easier for people understand what differences to expect.

File: src/ng/http.js [hunk 4]
```diff
@@ -122,8 +122,12 @@ function $HttpParamSerializerJQLikeProvider() {
...
+            serialize(toSerialize(), prefix);
```
Comment (gkalpak) [inline src/ng/http.js]: Actually, this is not how jQuery does it. They do not recursively serialize a returned object. They just [use the returned value][1].

[1]: https://github.com/jquery/jquery/blob/a6b0705294d336ae2f63f7276de0da1195495363/src/serialize.js#L60-L69

**Comment by frederikprijck** [comment]
Updated the PR, added a test for a function in an array aswell!

File: src/ng/http.js [hunk 5]
```diff
@@ -121,9 +121,13 @@ function $HttpParamSerializerJQLikeProvider() {
...
                 (topLevel ? '' : ']'));
           });
+        } else if (isFunction(toSerialize)) {
+          var toSerializeFnValue = toSerialize();
+          parts.push(encodeUriQuery(prefix) + '=' +
+                (toSerializeFnValue == null ? '' : encodeUriQuery(serializeValue(toSerializeFnValue))));
         } else {
           parts.push(encodeUriQuery(prefix) + '=' +
-              (toSerialize == null ? '' : encodeUriQuery(serializeValue(toSerialize))));
+                (toSerialize == null ? '' : encodeUriQuery(serializeValue(toSerialize))));
```
Comment (gkalpak) [inline src/ng/http.js]: Wrong indentation :grin:

Comment (gkalpak) [inline src/ng/http.js]: Actually, you can avoid code duplication with:

```js
  ...
} else {
  if (isFunction(toSerialize)) {
    toSerialize = toSerialize();
  }
  parts.push(encodeUriQuery(prefix) + '=' +
      (toSerialize == null ? '' : encodeUriQuery(serializeValue(toSerialize))));
}
```

File: test/ng/httpSpec.js [hunk 6]
```diff
@@ -2364,17 +2364,45 @@ describe('$http param serializers', function() {
       expect(decodeURIComponent(jqrSer({a: 'b', foo: ['bar', 'baz']}))).toEqual('a=b&foo[]=bar&foo[]=baz');
     });
 
+    it('should serialize arrays with functions', function() {
+      expect(jqrSer({foo:[function() { return 'bar'; }]})).toEqual('foo%5B%5D=bar');
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: Missing space after `:`. :grin:
BTW, you an probably use `valueFn('bar')` to save some typing 😎

File: test/ng/httpSpec.js [hunk 7]
```diff
@@ -2364,17 +2364,45 @@ describe('$http param serializers', function() {
...
+    });
+
     it('should serialize objects by repeating param name with [key] suffix', function() {
       expect(jqrSer({a: 'b', foo: {'bar': 'barv', 'baz': 'bazv'}})).toEqual('a=b&foo%5Bbar%5D=barv&foo%5Bbaz%5D=bazv');
                                                                            //a=b&foo[bar]=barv&foo[baz]=bazv
     });
 
+    it('should serialize objects with function properties by repeating param name with [key] suffix', function() {
+      expect(jqrSer({a: function() { return 'b'; }})).toEqual(
+        'a=b');
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: Remove the _"by repeating param name with [key] suffix"_ part, which is irrelevant (here and below).
Weird line-splitting (here and below) 😕

File: test/ng/httpSpec.js [hunk 8]
```diff
@@ -2364,17 +2364,45 @@ describe('$http param serializers', function() {
...
+        //a=b
+    });
+
+    it('should serialize objects with function properties returning an object by repeating param name with [key] suffix', function() {
+      expect(jqrSer({a: function() { return { b: 'c' }; }})).toEqual(
+        'a=%7B%22b%22:%22c%22%7D');
+        //a={"b": "c"}
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: There is no space after `:`.

File: test/ng/httpSpec.js [hunk 9]
```diff
@@ -2364,17 +2364,45 @@ describe('$http param serializers', function() {
...
 
+    it('should serialize objects with function properties by repeating param name with [key] suffix', function() {
+      expect(jqrSer({a: function() { return 'b'; }})).toEqual(
+        'a=b');
+        //a=b
+    });
+
+    it('should serialize objects with function properties returning an object by repeating param name with [key] suffix', function() {
+      expect(jqrSer({a: function() { return { b: 'c' }; }})).toEqual(
+        'a=%7B%22b%22:%22c%22%7D');
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: This is not exactly how jQuery does it, but that is not related to this PR :grin:

File: test/ng/httpSpec.js [hunk 10]
```diff
@@ -2364,17 +2364,45 @@ describe('$http param serializers', function() {
...
+    it('should serialize nested objects with function properties by repeating param name with [key] suffix', function() {
+      expect(jqrSer({foo: {'bar': function() { return 'barv'; }}})).toEqual(
+        'foo%5Bbar%5D=barv');
+        //foo[bar]=barv
+    });
+
+    it('should serialize nested objects with function properties returning an object by repeating param name with [key] suffix', function() {
+      expect(jqrSer({foo: {'bar': function() { return { bav: 'barv' }; }}})).toEqual(
+        'foo%5Bbar%5D=%7B%22bav%22:%22barv%22%7D');
+        //foo[bar]={"bav": "barv"}
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: No space after `:`.

File: test/ng/httpSpec.js [hunk 11]
```diff
@@ -2364,17 +2364,45 @@ describe('$http param serializers', function() {
...
+    });
+
     it('should serialize nested objects by repeating param name with [key] suffix', function() {
       expect(jqrSer({a: ['b', {c: 'd'}], e: {f: 'g', 'h': ['i', 'j']}})).toEqual(
          'a%5B%5D=b&a%5B1%5D%5Bc%5D=d&e%5Bf%5D=g&e%5Bh%5D%5B%5D=i&e%5Bh%5D%5B%5D=j');
          //a[]=b&a[1][c]=d&e[f]=g&e[h][]=i&e[h][]=j
     });
 
+    it('should serialize nested objects with function properties by repeating param name with [key] suffix', function() {
+      expect(jqrSer({foo: {'bar': function() { return 'barv'; }}})).toEqual(
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: No need for quoting `bar` (here and below).

**Review by gkalpak — CHANGES_REQUESTED** [review]
Please add some test for array inside object (e.g. `{foo: {bar: [valueFn('baz')]}}`).

**Comment by frederikprijck** [comment]
Updated!

File: src/ng/http.js [hunk 12]
```diff
@@ -121,13 +121,12 @@ function $HttpParamSerializerJQLikeProvider() {
...
-          var toSerializeFnValue = toSerialize();
-          parts.push(encodeUriQuery(prefix) + '=' +
-                (toSerializeFnValue == null ? '' : encodeUriQuery(serializeValue(toSerializeFnValue))));
         } else {
+          if (isFunction(toSerialize)) {
+            toSerialize = toSerialize();
+          }
           parts.push(encodeUriQuery(prefix) + '=' +
-                (toSerialize == null ? '' : encodeUriQuery(serializeValue(toSerialize))));
+            (toSerialize == null ? '' : encodeUriQuery(serializeValue(toSerialize))));
```
Comment (gkalpak) [inline src/ng/http.js]: Wrong indentation. 😞

File: test/ng/httpSpec.js [hunk 13]
```diff
@@ -2391,16 +2387,16 @@ describe('$http param serializers', function() {
          //a[]=b&a[1][c]=d&e[f]=g&e[h][]=i&e[h][]=j
     });
 
-    it('should serialize nested objects with function properties by repeating param name with [key] suffix', function() {
-      expect(jqrSer({foo: {'bar': function() { return 'barv'; }}})).toEqual(
+    it('should serialize nested objects with function properties', function() {
+      expect(jqrSer({foo: {bar: valueFn('barv')}})).toEqual(
         'foo%5Bbar%5D=barv');
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: Weird line break (here and below) :confused:

File: test/ng/httpSpec.js [hunk 14]
```diff
@@ -2391,16 +2387,16 @@ describe('$http param serializers', function() {
...
         //foo[bar]=barv
     });
 
-    it('should serialize nested objects with function properties returning an object by repeating param name with [key] suffix', function() {
-      expect(jqrSer({foo: {'bar': function() { return { bav: 'barv' }; }}})).toEqual(
+    it('should serialize nested objects with function properties returning an object', function() {
+      expect(jqrSer({foo: {bar: valueFn({ bav: 'barv' })}})).toEqual(
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: Inconsistent whitespace after `{` and before `}` :grin:

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: test/ng/httpSpec.js [hunk 15]
```diff
@@ -2364,17 +2364,41 @@ describe('$http param serializers', function() {
       expect(decodeURIComponent(jqrSer({a: 'b', foo: ['bar', 'baz']}))).toEqual('a=b&foo[]=bar&foo[]=baz');
     });
 
+    it('should serialize arrays with functions', function() {
+      expect(jqrSer({foo: [valueFn('bar')]})).toEqual('foo%5B%5D=bar');
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: Where's the decoded comment (`foo[]=bar`)? :grin:

File: test/ng/httpSpec.js [hunk 16]
```diff
@@ -2364,17 +2364,41 @@ describe('$http param serializers', function() {
...
     it('should serialize objects by repeating param name with [key] suffix', function() {
       expect(jqrSer({a: 'b', foo: {'bar': 'barv', 'baz': 'bazv'}})).toEqual('a=b&foo%5Bbar%5D=barv&foo%5Bbaz%5D=bazv');
                                                                            //a=b&foo[bar]=barv&foo[baz]=bazv
     });
 
+    it('should serialize objects with function properties', function() {
+      expect(jqrSer({a: valueFn('b')})).toEqual('a=b');
+    });
+
+    it('should serialize objects with function properties returning an object by repeating param name with [key] suffix', function() {
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: Inaccurate description.

File: test/ng/httpSpec.js [hunk 17]
```diff
@@ -2364,17 +2364,41 @@ describe('$http param serializers', function() {
...
          'a%5B%5D=b&a%5B1%5D%5Bc%5D=d&e%5Bf%5D=g&e%5Bh%5D%5B%5D=i&e%5Bh%5D%5B%5D=j');
          //a[]=b&a[1][c]=d&e[f]=g&e[h][]=i&e[h][]=j
     });
 
+    it('should serialize nested objects with function properties', function() {
+      expect(jqrSer({foo: {bar: valueFn('barv')}})).toEqual('foo%5Bbar%5D=barv'); //foo[bar]=barv
+    });
+
+    it('should serialize nested objects with function properties returning an object', function() {
+      expect(jqrSer({foo: {bar: valueFn({ bav: 'barv'})}})).toEqual('foo%5Bbar%5D=%7B%22bav%22:%22barv%22%7D'); //foo[bar]={"bav":"barv"}
```
Comment (gkalpak) [inline test/ng/httpSpec.js]: Extra whitespace in `{ bav`. 😱

**Review by gkalpak — CHANGES_REQUESTED** [review]

**Review by gkalpak — APPROVED** [review]

**Comment by gkalpak** [comment]
Merging to master only, since it has already deviated from 1.6.x (not converting empty strings to `null`).

**Pull request closed without merging**